scripts/olmo2_modal_openai.py

```diff
@@ -0,0 +1,251 @@
+
+import os
+import modal
+
+MODEL_NAME = "allenai/OLMo-2-1124-13B-Instruct"
+MODEL_REVISION = "9b7f0b17b9c6d6f6a0c3d1bba2495294f704ac2d"
+MODEL_DIR = "/root/models/{MODEL_NAME}"
+
+N_GPU = 1
+GPU_CONFIG = modal.gpu.A100(size="80GB", count=N_GPU)
+
+APP_NAME = "OLMo-2-1124-13B-Instruct-openai"
+APP_LABEL = APP_NAME.lower()
+
+MINUTES = 60  # seconds
+HOURS = 60 * MINUTES
+
+# ## Download the model weights
+
+# Our approach here differs from Modal's vllm_inference.py in two key ways:
+#
+# First, obviously, substituting an OLMo 2 model for the one from their example,
+#
+# Second, in their example, the weights are manually uploaded to a modal.com
+# storage volume, which is then mounted as a directory by the inference container.
+# In our example below, we use huggingface_hub's snapshot_download  to download
+# the weights from HuggingFace directly into a local directory when building the
+# container image.
+
+def download_model_to_image(model_dir, model_name, model_revision):
+    from huggingface_hub import snapshot_download
+    from transformers.utils import move_cache
+
+    os.makedirs(model_dir, exist_ok=True)
+
+    snapshot_download(
+        model_name,
+        revision=model_revision,
+        local_dir=model_dir,
+    )
+    move_cache()
+
+# ## Set up the container image
+
+# Our first order of business is to define the environment our server will run in
+# the container image. (See https://modal.com/docs/guide/custom-container)
+
+# This differs from vllm_interface.py in two major ways: first, as of the time this
+# is being written, the OLMo 2 model architecture requires building vLLM and transformers
+# from github commits that are too recent to have tagged versions, requiring that they
+# be built. The nvidia/cuda base image, git and build-essential apt_install, and the
+# torch and ray pip_installs, are all required to support these vllm and transformer
+# builds that follow.
+#
+# Second, we call the download_model_to_image function here, to build the resulting local
+# directory with model weights into our image.
+#
+# The image build can take a long time the first time you run this script; in particular,
+# the vLLM build can take well over an hour just by itself. The good news is that, as
+# long as the image definition doesn't change, Modal will cache and reuse the image on
+# later runs without having to re-run the full image build for each new container instance.
+
+vllm_image = (
+    modal.Image.from_registry("nvidia/cuda:12.2.0-devel-ubuntu22.04", add_python="3.10")
+    .apt_install("git", "build-essential")
+    .pip_install(
+        "torch==2.4.0",
+        "ray==2.10.0",
+        "hf-transfer==0.1.6",
+        "huggingface_hub==0.23.4",
+    )
+    .pip_install(
+        "git+https://github.com/vllm-project/vllm.git@9db713a1dca7e1bc9b6ecf5303c63c7352c52a13",
+    )
+    .pip_install(
+        "git+https://github.com/huggingface/transformers.git@9121ab8fe87a296e57f9846e70153b1a3c555d75",
+    )
```

Comment:
Do we still need to depend on shas? I think both vllm and huggingface now have OLMo in main, if not in the latest released version.

Reply:
huggingface/transformers is on a tagged version that includes olmo2 as of 5 hours ago, but vLLM still requires a build. I'll try it with main instead of the specific sha, which should work, but that'll be an hour or two for the build to confirm it.

Reply:
:+1:

Reply:
After running this for a while, I'm going to swing back to advocating for pinning, either to a sha or to a tagged version once that catches up, because otherwise we have the likelihood of unacceptable delays to build new versions if we spin up new containers in response to a surge in traffic and a new version has come out since the last time we brought up the container.

With that in mind, do we want to go with a pinned sha we've got, or wait until we get a tagged vllm version with the olmo2 architecture, which I'm guessing is mid-month?